compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt

```diff
@@ -0,0 +1,95 @@
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.diagnostics.ConeSimpleDiagnostic
+import org.jetbrains.kotlin.fir.diagnostics.DiagnosticKind
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.expressions.arguments
+import org.jetbrains.kotlin.fir.references.FirErrorNamedReference
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeAmbiguityError
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeInapplicableCandidateError
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeUnresolvedNameError
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.fir.visitors.FirVisitorVoid
+import org.jetbrains.kotlin.resolve.calls.tower.isSuccess
+
+object FirDelegatedPropertyChecker : FirPropertyChecker() {
+    override fun check(declaration: FirProperty, context: CheckerContext, reporter: DiagnosticReporter) {
+        val delegate = declaration.delegate ?: return
+        val delegateType = delegate.typeRef.coneTypeSafe<ConeKotlinType>() ?: return
+
+        if (delegateType is ConeKotlinErrorType) {
+            val delegateSource = delegate.source
+            // Implicit recursion type is not reported since the type ref does not have a real source.
+            if (delegateSource != null && (delegateType.diagnostic as? ConeSimpleDiagnostic)?.kind == DiagnosticKind.RecursionInImplicitTypes) {
+                // skip reporting other issues in this case
+                reporter.reportOn(delegateSource, FirErrors.RECURSION_IN_IMPLICIT_TYPES, context)
+                return
+            }
+        }
+
+        class Visitor(private val isGet: Boolean) : FirVisitorVoid() {
+            override fun visitElement(element: FirElement) = element.acceptChildren(this)
+
+            override fun visitFunctionCall(functionCall: FirFunctionCall) {
+                val errorNamedReference = functionCall.calleeReference as? FirErrorNamedReference ?: return
+                if (errorNamedReference.source?.kind != FirFakeSourceElementKind.DelegatedPropertyAccessor) return
+                val calleeArgTypes = "(${functionCall.arguments.joinToString(", ") { it.typeRef.coneType.render() }})"
+                when (val diagnostic = errorNamedReference.diagnostic) {
+                    is ConeUnresolvedNameError -> {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_MISSING,
+                            "${diagnostic.name}$calleeArgTypes",
+                            delegateType,
+                            if (isGet) "delegate" else "delegate for var (read-write property)",
+                            context
+                        )
+                    }
+                    is ConeAmbiguityError -> {
+                        if (diagnostic.applicability.isSuccess) {
```

Comment:
Well, this is read funny, like it is success, but error! (kidding).

Maybe add a comment, like `// applicable yet ambiguous, i.e., multiple candidates`